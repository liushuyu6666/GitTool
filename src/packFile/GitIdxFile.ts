import { readFileSync } from 'fs';
import path from 'path';
import { GitFanout } from './GitFanout';

export interface GitIdxFileInterface {
  byteSize: number;

  header: number[];

  versionNumber: number;

  fanout: GitFanout;

  parseHeader(content: Buffer): [number, number, number, number];

  parseVersionNumber(content: Buffer): number

}

export class GitIdxFile implements GitIdxFileInterface {
  // size by bytes
  byteSize: number;

  header: number[];

  versionNumber: number;

  fanout: GitFanout;

  constructor(filePath: string, outDir: string) {
    const fileName = filePath.substring(filePath.lastIndexOf(path.sep));
    const outFile = path.join(outDir, fileName);
    const content = readFileSync(filePath);
    this.byteSize = content.length;
    this.header = this.parseHeader(content);
    this.versionNumber = this.parseVersionNumber(content);
    this.fanout = new GitFanout(content, outFile);
  }

  parseHeader(content: Buffer): [number, number, number, number] {
    // [0, 3] bytes
    return [
      content.readUInt8(0),
      content.readUInt8(1),
      content.readUInt8(2),
      content.readUInt8(3),
    ];
  }

  parseVersionNumber(content: Buffer): number {
    return content.readUInt32BE(4);
  }
}
