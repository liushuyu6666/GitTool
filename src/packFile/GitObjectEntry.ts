import { ManipulateBuffer } from "../ManipulateBuffer/ManipulateBuffer";

export interface GitObjectEntryInterface {
  type: number;

  variableLengthInteger: number;

  content: Buffer;

  getAndParseVariableLengthInteger(chunk: Buffer): [number, number];
}

export class GitObjectEntry implements GitObjectEntryInterface {
  
  type: number;

  variableLengthInteger: number;

  content: Buffer;

  constructor(chunk: Buffer) {
    this.type = (chunk[0] & 0b01110000) >> 4 & 0b00000111;

    let contentStartIndex = 0;

    [this.variableLengthInteger, contentStartIndex] = this.getAndParseVariableLengthInteger(chunk);

    this.content = chunk.subarray(contentStartIndex);
  }

  getAndParseVariableLengthInteger(chunk: Buffer): [number, number] {
    const manipulateBuffer = new ManipulateBuffer(false);
    let msb: boolean = chunk[0] > 0b10000000;
    let index = 1;
    manipulateBuffer.fill(chunk[0], 4, 8);
    while (msb) {
      manipulateBuffer.fill(chunk[index], 1, 8);
      msb = chunk[index] > 0b10000000
      index++;
    }
    manipulateBuffer.finish();

    return [manipulateBuffer.readIntBE(), index];
  }
}
