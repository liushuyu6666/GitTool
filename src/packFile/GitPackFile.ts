import { readFileSync } from 'fs';
import { Offset } from './GitFanout';
import { GitObjectEntry } from './GitObjectEntry';

// export interface PackInfo {
//   type: number;
//   info: string;
// }

export interface GitPackInterfaceFile {
  size: number;

  // 'PACK'
  layer1: string;

  // version
  layer2: number;

  // object hex list
  layer3: number;

  // data chunk
  layer4: Record<string, GitObjectEntry>;

  // checksum
  layer5: Buffer;

  // layer 1: [0, 4)
  parseLayer1(content: Buffer): string;
  // layer 2: [4, 8)
  parseLayer2(content: Buffer): number;

  // layer 3: [8, 12)
  parseLayer3(content: Buffer): number;

  parseLayer4(offsets: Record<string, number>, content: Buffer): Record<string, GitObjectEntry>;

  parseLayer5(content: Buffer): Buffer;
}

// TODO: need setter and getter
export class GitPackFile implements GitPackInterfaceFile {
  size: number;

  // 'PACK'
  layer1: string;

  // version
  layer2: number;

  // object hex list
  layer3: number;

  // data chunk
  layer4: Record<string, GitObjectEntry>;

  // checksum
  layer5: Buffer;

  constructor(filePath: string, offsets: Record<string, number>) {
    const content = readFileSync(filePath);
    this.size = content.length;
    this.layer1 = this.parseLayer1(content);
    this.layer2 = this.parseLayer2(content);
    this.layer3 = this.parseLayer3(content);
    this.layer4 = this.parseLayer4(offsets, content);
    this.layer5 = this.parseLayer5(content);
  }

  // layer 1: [0, 4)
  parseLayer1(content: Buffer): string {
    return content.subarray(0, 4).toString('utf-8');
  }

  // layer 2: [4, 8)
  parseLayer2(content: Buffer): number {
    return content.subarray(4, 8).readUInt32BE();
  }

  // layer 3: [8, 12)
  parseLayer3(content: Buffer): number {
    return content.subarray(8, 12).readUInt32BE();
  }

  parseLayer4(offsets: Record<string, number>, content: Buffer): Record<string, GitObjectEntry> {
    // create offsetArray
    const offsetArray: Offset[] = [];
    Object.entries(offsets).forEach(([hex, offset]) => {
      offsetArray.push({
        hex,
        offset
      })
    })

    // sort offsetArray ascending
    // offsetArray.length = offsets.length + 1
    offsetArray.sort((prev, next) => prev.offset - next.offset);
    offsetArray.push({
      hex: '',
      offset: content.length - 20 // this is the endIndex of the last object entry
    })

    const gitPackObjectEntry: Record<string, GitObjectEntry> = {};
    for (let i = 0; i < Object.keys(offsets).length; i++) {
      const startIndex = offsetArray[i].offset;
      const endIndex = offsetArray[i + 1].offset;
      const hex = offsetArray[i].hex;
      const entry = new GitObjectEntry(content, startIndex, endIndex);
      gitPackObjectEntry[hex] = entry;
    }
    return gitPackObjectEntry;
  }

  parseLayer5(content: Buffer): Buffer {
    return content.subarray(content.length - 20);
  }
}
