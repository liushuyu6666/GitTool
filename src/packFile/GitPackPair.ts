import { GitObject } from '../gitObject/GitObject';
import { isPackDeltaObject, isPackOriginalObject } from '../utils/getGitObjectType';
import { GitIdxFile } from './GitIdxFile';
import { GitPackFile } from './GitPackFile';

export interface GitPackPairInterface {
  pack: GitPackFile;

  idx: GitIdxFile;
}

export interface GitPackObjects {
  packOriginalObjects: GitObject[];
  packDeltaObjects: GitObject[];
}

export class GitPackPair implements GitPackPairInterface {
  pack: GitPackFile;

  idx: GitIdxFile;

  filePathIdx: string;

  filePathPack: string;

  constructor(filePathIdx: string, filePathPack: string, outPackPath: string) {
    this.filePathIdx = filePathIdx;
    this.filePathPack = filePathPack;
    this.idx = new GitIdxFile(filePathIdx, outPackPath);
    this.pack = new GitPackFile(filePathPack, this.idx.fanout.offsets);
  }

  generateGitPackObjects(): GitPackObjects {
    const packOriginalObjects: GitObject[] = [];
    const packDeltaObjects: GitObject[] = [];

    for (const [hash, gitObjectEntry] of Object.entries(this.pack.layer4)) {
      const type = gitObjectEntry.type;

      if (isPackOriginalObject(type)) {
        packOriginalObjects.push(
          new GitObject({
            hash,
            type,
            size: gitObjectEntry.size,
            filePath: this.filePathPack,
            bodyOffsetStartIndex: gitObjectEntry.bodyStartIndex,
            bodyOffsetEndIndex: gitObjectEntry.bodyEndIndex,
          }),
        );
      } else if (isPackDeltaObject(type)) {
        packDeltaObjects.push(
          new GitObject({
            hash,
            type,
            size: gitObjectEntry.size,
            filePath: this.filePathPack,
            bodyOffsetStartIndex: gitObjectEntry.bodyStartIndex,
            bodyOffsetEndIndex: gitObjectEntry.bodyEndIndex,
          }),
        );
      }
    }

    return {
      packOriginalObjects,
      packDeltaObjects,
    };
  }

  // generateGitOriginalObject(): GitObject[] {
  //   const gitObjects: GitObject[] = [];

  //   for (const [hash, gitObjectEntry] of Object.entries(this.pack.layer4)) {
  //     const type = gitObjectEntry.type;

  //     if (isOriginalDeltaObject(type)) {
  //       gitObjects.push(
  //         new GitObject({
  //           hash,
  //           type,
  //           size: gitObjectEntry.size,
  //           filePath: this.filePathPack,
  //           bodyOffsetStartIndex: gitObjectEntry.bodyStartIndex,
  //           bodyOffsetEndIndex: gitObjectEntry.bodyEndIndex,
  //         }),
  //       );
  //     }
  //   }

  //   return gitObjects;
  // }

  // TODO: maybe we can combine generateGitDeltaObject with generateGitOriginalObject
  // generateGitDeltaObject(): GitObject[] {
  //   const gitObjects: GitObject[] = [];

  //   for (const [hash, gitObjectEntry] of Object.entries(this.pack.layer4)) {
  //     const type = gitObjectEntry.type;

  //     if (isDeltaObject(type)) {
  //       gitObjects.push(
  //         new GitObject({
  //           hash,
  //           type,
  //           size: gitObjectEntry.size,
  //           filePath: this.filePathPack,
  //           bodyOffsetStartIndex: gitObjectEntry.bodyStartIndex,
  //           bodyOffsetEndIndex: gitObjectEntry.bodyEndIndex,
  //         }),
  //       );
  //     }
  //   }

  //   return gitObjects;
  // }
}
