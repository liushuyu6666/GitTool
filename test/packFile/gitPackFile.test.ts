import { inflateSync } from "zlib";
import { GitIdxFile } from "../../src/packFile/GitIdxFile";
import { GitPackFile } from "../../src/packFile/GitPackFile";

describe('test pack', () => {
  const filePath =
    'gitTest/objects/pack/pack-5eb139577548314dffcb8e6410b30413a81ca3fb.pack';
  const idxFilePath = 
    'gitTest/objects/pack/pack-5eb139577548314dffcb8e6410b30413a81ca3fb.idx';
  const gitIdx = new GitIdxFile(idxFilePath);
  const gitPack = new GitPackFile(filePath, gitIdx.fanout.offsets);

  test('test layer1, should be PACK', () => {
    expect(gitPack.layer1).toBe('PACK');
  });

  test('test data chunk', () => {
    const layer4 = gitPack.layer4;
    expect(Object.keys(layer4)).toMatchSnapshot('packObjectEntries');
    expect(Object.keys(layer4).length).toBe(167);
  })

  test('check OBJ_COMMIT', () => {
    const layer4 = gitPack.layer4;
    const buffer: Buffer[] = [];
    Object.entries(layer4).forEach(([_, entry]) => {
      // When type is OBJ_COMMIT
      if (entry.type === 1) {
        buffer.push(inflateSync(entry.content));
      }
    })

    expect(buffer).toMatchSnapshot('OBJ_COMMIT')
  })
});