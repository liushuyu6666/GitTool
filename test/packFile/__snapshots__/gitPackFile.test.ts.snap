// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test pack test data chunk: packObjectEntries 1`] = `
[
  "c17358365df785db682260407606051b011eee88",
  "3b18e512dba79e4c8300dd08aeb37f8e728b8dad",
  "c3b8bb102afeca86037d5b5dd89ceeb0090eae9d",
]
`;
